### PR TITLE
Improvements for error reporting for Subscription updates / 401s

### PR DESCRIPTION
**Pull request by MattGal**
Addresses 
- https://github.com/dotnet/core-eng/issues/10066
- https://github.com/dotnet/core-eng/issues/10076
- Issue that it itself reported: https://github.com/maestro-auth-test/maestro-test2/issues/11267

### CODE REVIEW

File: src/Maestro/DependencyUpdateErrorProcessor/DependencyUpdateErrorProcessor.cs [hunk 1]
```diff
@@ -69,14 +70,14 @@ public async Task ProcessDependencyUpdateErrorsAsync()
...
                     using (ITransaction tx = _stateManager.CreateTransaction())
                     {
-                         checkpoint = await checkpointEvaluator.GetOrAddAsync(
-                         tx,
-                         "checkpointEvaluator",
-                         DateTimeOffset.UtcNow
+                        checkpoint = await checkpointEvaluator.GetOrAddAsync(
+                        tx,
+                        "checkpointEvaluator",
+                        DateTimeOffset.UtcNow
```
Comment (riarenas) [inline src/Maestro/DependencyUpdateErrorProcessor/DependencyUpdateErrorProcessor.cs:76]: [nit] I think the `)` should also lose a space.

File: src/Maestro/DependencyUpdateErrorProcessor/DependencyUpdateErrorProcessor.cs [hunk 2]
```diff
@@ -151,54 +161,85 @@ public async Task ProcessDependencyUpdateErrorsAsync()
...
-            RepositoryBranchUpdateHistoryEntry updateHistoryError, 
+            UpdateHistoryEntry updateHistoryError, 
             string issueRepo, 
             Func<string, string, bool> shouldReplaceDescription, 
             string description)
         {
-            _logger.LogInformation($"Error Message : '{updateHistoryError.ErrorMessage}' in repository :  '{updateHistoryError.Repository}'");
-            IReliableDictionary<(string repository, string branch),int> gitHubIssueEvaluator =
-                await _stateManager.GetOrAddAsync<IReliableDictionary<(string repository, string branch), int>>("gitHubIssueEvaluator");
+
```
Comment (riarenas) [inline src/Maestro/DependencyUpdateErrorProcessor/DependencyUpdateErrorProcessor.cs:169]: [nit] extra blank line

File: src/Maestro/DependencyUpdateErrorProcessor/DependencyUpdateErrorProcessor.cs [hunk 3]
```diff
@@ -151,54 +161,85 @@ public async Task ProcessDependencyUpdateErrorsAsync()
...
             IGitHubClient client = await _authenticateGitHubApplicationClient.CreateGitHubClientAsync(parseRepoUri.owner, parseRepoUri.repo);
-            Octokit.Repository repo = await client.Repository.Get(
-                parseRepoUri.owner, 
+            Repository repo = await client.Repository.Get(
+                parseRepoUri.owner,
                 parseRepoUri.repo);
             var issueNumber = new ConditionalValue<int>();
-            using (ITransaction tx = _stateManager.CreateTransaction())
+
+            if (updateHistoryError is RepositoryBranchUpdateHistoryEntry)
```
Comment (riarenas) [inline src/Maestro/DependencyUpdateErrorProcessor/DependencyUpdateErrorProcessor.cs:177]: How do you feel about using a pattern matching switch here? https://docs.microsoft.com/en-us/dotnet/csharp/pattern-matching#using-pattern-matching-switch-statements

something like:

```C#
switch (updateHistoryError)
{
  case RepositoryBranchUpdateHistoryEntry repoBranchUpdateHistoryError:
    blah
    break;
  case SubscriptionUpdateHistoryEntry subscriptionUpdateHistoryError  :
    blah;
    break
  default: 
   well.. this is awkward
   break;
}
```

File: src/Maestro/DependencyUpdateErrorProcessor/DependencyUpdateErrorProcessor.cs [hunk 4]
```diff
@@ -151,54 +161,85 @@ public async Task ProcessDependencyUpdateErrorsAsync()
...
+                using (ITransaction tx = _stateManager.CreateTransaction())
+                {
+                    issueNumber = await gitHubIssueEvaluator.TryGetValueAsync(
+                        tx,
+                        (repoBranchUpdateHistoryError.Repository,
+                         repoBranchUpdateHistoryError.Branch));
+                    await tx.CommitAsync();
+                }
+
+                // Create a new issue for the error if the issue is already closed or the issue does not exists.
```
Comment (riarenas) [inline src/Maestro/DependencyUpdateErrorProcessor/DependencyUpdateErrorProcessor.cs:194]:
```suggestion
                // Create a new issue for the error if the issue is already closed or the issue does not exist.
```

File: src/Maestro/DependencyUpdateErrorProcessor/DependencyUpdateErrorProcessor.cs [hunk 5]
```diff
@@ -257,7 +311,48 @@ public static (string owner, string repo) ParseRepoUri(string uri)
                     (key, value) => issue.Number);
                 await tx.CommitAsync();
             }
+        }
+
+
```
Comment (riarenas) [inline src/Maestro/DependencyUpdateErrorProcessor/DependencyUpdateErrorProcessor.cs:316]: [nit] double blank line

File: src/Maestro/DependencyUpdateErrorProcessor/DependencyUpdateErrorProcessor.cs [hunk 6]
```diff
@@ -257,7 +311,48 @@ public static (string owner, string repo) ParseRepoUri(string uri)
...
+        /// <summary>
+        /// Creates a new github issue.
+        /// </summary>
+        /// <param name="client">Github client</param>
+        /// <param name="updateHistoryError">Error info for which github issue has to be created</param>
+        /// <param name="gitHubIssueEvaluator">Reliable service that holds the (key - repo-branch) and (value issueNumber)</param>
```
Comment (riarenas) [inline src/Maestro/DependencyUpdateErrorProcessor/DependencyUpdateErrorProcessor.cs:322]:
```suggestion
        /// <param name="gitHubIssueEvaluator">Reliable collection that holds the (key - repo-branch) and (value issueNumber)</param>
```

**Comment by riarenas** [comment]
The changes look good to me overall. Thanks Matt!. The example issue you linked is a little hard to read. Are there some formatting changes we need to make so the markdown is happy?

File: src/Maestro/DependencyUpdateErrorProcessor/DependencyUpdateErrorProcessor.cs (hunk 3 above)
Comment (MattGal) [reply src/Maestro/DependencyUpdateErrorProcessor/DependencyUpdateErrorProcessor.cs:177]: that was just not in my arsenal, sure.

**Comment by epananth** [comment]
Changes look good to me, I especially like the switch statement you added to identify which table the error is originating. 

Only one q, the error you linked in the issue, did you edit the marker tag so that we can see what is in it? Cos normally it is hidden.

**Comment by MattGal** [comment]
> Changes look good to me, I especially like the switch statement you added to identify which table the error is originating.
> 
> Only one q, the error you linked in the issue, did you edit the marker tag so that we can see what is in it? Cos normally it is hidden.

Hrm I did not, what normally hides it?  Can you link to a good example?

**Pull request merged**